File: qmd-python/src/qmd/formatter/formatter.py
"""Output formatting for search results"""

import csv
import json
import io
from enum import Enum
from typing import Optional

from qmd.store.store import SearchResult


class Format(str, Enum):
    """Output format types"""
    CLI = "cli"
    JSON = "json"
    MARKDOWN = "markdown"
    CSV = "csv"
    FILES = "files"
    XML = "xml"


class Formatter:
    """Formats search results for output"""

    def __init__(self, format_str: str = "cli", limit: int = 20):
        self.format = Format(format_str) if format_str else Format.CLI
        self.limit = limit

    def format_results(self, results: list[SearchResult], query: str = "") -> str:
        """Format search results"""
        if len(results) > self.limit:
            results = results[:self.limit]

        match self.format:
            case Format.JSON:
                return self._format_json(results, query)
            case Format.MARKDOWN:
                return self._format_markdown(results)
            case Format.CSV:
                return self._format_csv(results)
            case Format.FILES:
                return self._format_files(results)
            case Format.XML:
                return self._format_xml(results)
            case _:
                return self._format_cli(results)

    def _format_cli(self, results: list[SearchResult]) -> str:
        """Format as CLI table"""
        if not results:
            return "No results found.\n"

        lines = []
        lines.append(f"{'Score':<10} {'Lines':<8} {'DocID':<6} {'Path'}")
        lines.append("-" * 60)

        for r in results:
            lines.append(f"{r.score:<10.4f} {r.lines:<8} {r.doc_id:<6} {r.path}")

        lines.append(f"\nTotal: {len(results)} results\n")
        return "\n".join(lines)

    def _format_json(self, results: list[SearchResult], query: str) -> str:
        """Format as JSON"""
        output = {
            "query": query,
            "total": len(results),
            "results": [
                {
                    "doc_id": r.doc_id,
                    "path": r.path,
                    "collection": r.collection,
                    "score": r.score,
                    "lines": r.lines,
                    "title": r.title,
                    "hash": r.hash,
                }
                for r in results
            ],
        }
        return json.dumps(output, indent=2) + "\n"

    def _format_markdown(self, results: list[SearchResult]) -> str:
        """Format as Markdown"""
        lines = ["# Search Results\n", f"Total: {len(results)} results\n"]

        for i, r in enumerate(results):
            lines.append(f"## {i + 1}. {r.path}\n")
            lines.append(f"- **Score**: {r.score:.4f}")
            lines.append(f"- **Lines**: {r.lines}")
            lines.append(f"- **Collection**: {r.collection}")
            lines.append(f"- **DocID**: {r.doc_id}\n")

        return "\n".join(lines)

    def _format_csv(self, results: list[SearchResult]) -> str:
        """Format as CSV"""
        output = io.StringIO()
        writer = csv.writer(output)
        writer.writerow(["score", "lines", "docid", "path", "collection", "title", "hash"])

        for r in results:
            writer.writerow([
                f"{r.score:.4f}",
                r.lines,
                r.doc_id,
                r.path,
                r.collection,
                r.title,
                r.hash,
            ])

        return output.getvalue()

    def _format_files(self, results: list[SearchResult]) -> str:
        """Format as file list"""
        return "\n".join(r.path for r in results) + "\n"

    def _format_xml(self, results: list[SearchResult]) -> str:
        """Format as XML"""
        lines = ['<?xml version="1.0" encoding="UTF-8"?>', "<results>"]
        lines.append(f"  <total>{len(results)}</total>")

        for r in results:
            lines.append("  <result>")
            lines.append(f"    <score>{r.score:.4f}</score>")
            lines.append(f"    <lines>{r.lines}</lines>")
            lines.append(f"    <docid>{self._escape_xml(r.doc_id)}</docid>")
            lines.append(f"    <path>{self._escape_xml(r.path)}</path>")
            lines.append(f"    <collection>{self._escape_xml(r.collection)}</collection>")
            lines.append(f"    <title>{self._escape_xml(r.title)}</title>")
            lines.append(f"    <hash>{r.hash}</hash>")
            lines.append("  </result>")

        lines.append("</results>\n")
        return "\n".join(lines)

    @staticmethod
    def _escape_xml(s: str) -> str:
        """Escape XML special characters"""
        return (
            s.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
            .replace('"', "&quot;")
            .replace("'", "&apos;")
        )

File: qmd-go/pkg/formatter/formatter.go
package formatter

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"strings"

	"github.com/qmd/qmd-go/pkg/store"
)

// Format type
type Format string

const (
	FormatCLI     Format = "cli"
	FormatJSON    Format = "json"
	FormatMarkdown Format = "markdown"
	FormatCSV     Format = "csv"
	FormatFiles   Format = "files"
	FormatXML     Format = "xml"
)

// Formatter formats search results
type Formatter struct {
	format Format
	limit  int
}

// New creates a new Formatter
func New(format string, limit int) *Formatter {
	if limit == 0 {
		limit = 20
	}

	f := Format(format)
	if f == "" {
		f = FormatCLI
	}

	return &Formatter{
		format: f,
		limit:  limit,
	}
}

// FormatSearchResults formats search results
func (f *Formatter) FormatSearchResults(results []store.SearchResult) string {
	if len(results) > f.limit {
		results = results[:f.limit]
	}

	switch f.format {
	case FormatJSON:
		return f.formatJSON(results)
	case FormatMarkdown:
		return f.formatMarkdown(results)
	case FormatCSV:
		return f.formatCSV(results)
	case FormatFiles:
		return f.formatFiles(results)
	case FormatXML:
		return f.formatXML(results)
	default:
		return f.formatCLI(results)
	}
}

func (f *Formatter) formatCLI(results []store.SearchResult) string {
	if len(results) == 0 {
		return "No results found.\n"
	}

	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("%-10s %-8s %-6s %s\n", "Score", "Lines", "DocID", "Path"))
	sb.WriteString(fmt.Sprintf("%-10s %-8s %-6s %s\n", "------", "-----", "-----", "----"))

	for _, r := range results {
		sb.WriteString(fmt.Sprintf("%-10.4f %-8d %-6d %s\n", r.Score, r.Lines, r.DocID, r.Path))
	}

	sb.WriteString(fmt.Sprintf("\nTotal: %d results\n", len(results)))
	return sb.String()
}

func (f *Formatter) formatJSON(results []store.SearchResult) string {
	type Output struct {
		Query  string            `json:"query"`
		Total  int               `json:"total"`
		Results []store.SearchResult `json:"results"`
	}

	output := Output{
		Query:   "",
		Total:   len(results),
		Results: results,
	}

	data, _ := json.MarshalIndent(output, "", "  ")
	return string(data) + "\n"
}

func (f *Formatter) formatMarkdown(results []store.SearchResult) string {
	var sb strings.Builder
	sb.WriteString("# Search Results\n\n")
	sb.WriteString(fmt.Sprintf("Total: %d results\n\n", len(results)))

	for i, r := range results {
		sb.WriteString(fmt.Sprintf("## %d. %s\n\n", i+1, r.Path))
		sb.WriteString(fmt.Sprintf("- **Score**: %.4f\n", r.Score))
		sb.WriteString(fmt.Sprintf("- **Lines**: %d\n", r.Lines))
		sb.WriteString(fmt.Sprintf("- **Collection**: %s\n", r.Collection))
		sb.WriteString(fmt.Sprintf("- **DocID**: %s\n\n", r.DocID))
	}

	return sb.String()
}

func (f *Formatter) formatCSV(results []store.SearchResult) string {
	var sb strings.Builder
	writer := csv.NewWriter(&sb)
	defer writer.Flush()

	// Header
	writer.Write([]string{"score", "lines", "docid", "path", "collection", "title", "hash"})

	// Data
	for _, r := range results {
		writer.Write([]string{
			fmt.Sprintf("%.4f", r.Score),
			fmt.Sprintf("%d", r.Lines),
			r.DocID,
			r.Path,
			r.Collection,
			r.Title,
			r.Hash,
		})
	}

	return sb.String()
}

func (f *Formatter) formatFiles(results []store.SearchResult) string {
	var sb strings.Builder
	for _, r := range results {
		sb.WriteString(r.Path)
		sb.WriteString("\n")
	}
	return sb.String()
}

func (f *Formatter) formatXML(results []store.SearchResult) string {
	var sb strings.Builder
	sb.WriteString("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n")
	sb.WriteString("<results>\n")
	sb.WriteString(fmt.Sprintf("  <total>%d</total>\n", len(results)))

	for _, r := range results {
		sb.WriteString("  <result>\n")
		sb.WriteString(fmt.Sprintf("    <score>%.4f</score>\n", r.Score))
		sb.WriteString(fmt.Sprintf("    <lines>%d</lines>\n", r.Lines))
		sb.WriteString(fmt.Sprintf("    <docid>%s</docid>\n", r.DocID))
		sb.WriteString(fmt.Sprintf("    <path>%s</path>\n", escapeXML(r.Path)))
		sb.WriteString(fmt.Sprintf("    <collection>%s</collection>\n", r.Collection))
		sb.WriteString(fmt.Sprintf("    <title>%s</title>\n", escapeXML(r.Title)))
		sb.WriteString(fmt.Sprintf("    <hash>%s</hash>\n", r.Hash))
		sb.WriteString("  </result>\n")
	}

	sb.WriteString("</results>\n")
	return sb.String()
}

func escapeXML(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	s = strings.ReplaceAll(s, ">", "&gt;")
	s = strings.ReplaceAll(s, "\"", "&quot;")
	s = strings.ReplaceAll(s, "'", "&apos;")
	return s
}
